### PR TITLE
fix: do not watch when karma autoWatch is set to false

### PR DESCRIPTION
**Pull request by mjeanroy**
Hi,

When I use the `@next` version of `karma-rollup-preprocessor` (so 7.0.0-rc2 for now), it looks like my unit test suite never ends when `autoWatch` is set to `false`.

I tried to find the reason, and it looks like this is due to this [commit](https://github.com/jlmakes/karma-rollup-preprocessor/commit/3d9058b6878e55babeed769cc037e4fbff5bc6bf). I think the main reason is because files to watch are never removed from the internal map.

This fix may not be the most appropriate, but at least it add a little improvement and do not track files to watch if it is not really needed (so if `autoWatch` is `false`), and fix the issue mentioned above.

### CODE REVIEW

**Comment by jlmakes** [comment]
Hey Mickael

Thanks for your contributions, let’s see if we can get to the bottom of this! If you don't mind, I'd like to first share my understanding of the Karma source (as it pertains to `config.autoWatch`) — I think this will help us communicate about your problem.

Most of Karma is wrapped up into the `Server` found at `./lib/server.js`, and a search for `config.autoWatch` has three matches (all in the `_start` method):

<details>
<summary><i>Expand for code review</i></summary>

***

[First](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L168-L170) — It is used to initialize the Karma watcher after preprocessing completes:

```js
const afterPreprocess = () => {
  if (config.autoWatch) {
    this._injector.invoke(watcher.watch);
  }
  // other code...
};
```

This doesn’t really impact us, because we should be passing `watched: false` to the Karma config:

```js
files: [
  { pattern: 'test/**/*.spec.js', watched: false }
],
```

***

[Second](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L191-L201) — It is used for the first run of tests after all browsers are ready:

```js
this.on("browser_register", browser => {
  launcher.markCaptured(browser.id);

  if (launcher.areAllCaptured()) {
    this.emit("browsers_ready");

    if (config.autoWatch) {
      executor.schedule();
    }
  }
});

```

***

[Third](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L324-L332) — It is used to bind `executor.schedule()` to the `"file_list_modified"` event:

```js
if (config.autoWatch) {
  this.on("file_list_modified", () => {
    this.log.debug("List of files has changed, trying to execute");
    if (config.restartOnFileChange) {
      socketServer.sockets.emit("stop");
    }
    executor.schedule();
  });
}
```

Both the Karma’s watcher and the `karma-rollup-preprocessor@next` watcher rely upon this event handler to re-run tests.

</details>

***

We can see that by setting `config.autoWatch` to `false`, we are telling Karma:

- Don‘t initialize a watcher
- Don‘t run tests after `karma start`
- Don’t re-run tests when files are modified

Which means your test suite will just hang: no tests will run, even if you modify files. You can _force_ your tests to run in this situation using `config.singleRun == true`, which under the hood emits the same `"run_start"` event that `executor.schedule()` does.

> ...it looks like my unit test suite never ends when autoWatch is set to false.

**Either`config.singleRun` or `config.autoWatch` must be `true` in order for your tests to run.**

The only exception to this would be if you are running Karma programmatically (not from the command line), in which case you could manually invoke `executor.schedule()` or emit your own `"run_start"` events.

***

**TL;DR**: based on the way Karma is designed, it looks like we almost always want `config.autoWatch` set to `true` — and we simply toggle `config.singleRun` as needed.

Does any of this help you?

**Comment by jlmakes** [comment]
Hey Mickael

Thanks for your contributions, let’s see if we can get to the bottom of this! If you don't mind, I'd like to first share my understanding of the Karma source (as it pertains to `config.autoWatch`) — I think this will help us communicate about your problem.

Most of Karma is wrapped up into the `Server` found at `./lib/server.js`, and a search for `config.autoWatch` has three matches (all in the `_start` method):

<details>
<summary><i>Expand for code review</i></summary>

***

[First](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L168-L170) — It is used to initialize the Karma watcher after preprocessing completes:

```js
const afterPreprocess = () => {
  if (config.autoWatch) {
    this._injector.invoke(watcher.watch);
  }
  // other code...
};
```

This doesn’t really impact us, because we should be passing `watched: false` to the Karma config:

```js
files: [
  { pattern: 'test/**/*.spec.js', watched: false }
],
```

***

[Second](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L191-L201) — It is used for the first run of tests after all browsers are ready:

```js
this.on("browser_register", browser => {
  launcher.markCaptured(browser.id);

  if (launcher.areAllCaptured()) {
    this.emit("browsers_ready");

    if (config.autoWatch) {
      executor.schedule();
    }
  }
});

```

***

[Third](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L324-L332) — It is used to bind `executor.schedule()` to the `"file_list_modified"` event:

```js
if (config.autoWatch) {
  this.on("file_list_modified", () => {
    this.log.debug("List of files has changed, trying to execute");
    if (config.restartOnFileChange) {
      socketServer.sockets.emit("stop");
    }
    executor.schedule();
  });
}
```

Both the Karma’s watcher and the `karma-rollup-preprocessor@next` watcher rely upon this event handler to re-run tests.

</details>

***

We can see that by setting `config.autoWatch` to `false`, we are telling Karma:

- Don‘t initialize a watcher
- Don‘t run tests after `karma start`
- Don’t re-run tests when files are modified

Which means your test suite will just hang: no tests will run, even if you modify files. You can _force_ your tests to run in this situation using `config.singleRun == true`, which under the hood emits the same `"run_start"` event that `executor.schedule()` does.

> ...it looks like my unit test suite never ends when autoWatch is set to false.

**Either`config.singleRun` or `config.autoWatch` must be `true` in order for your tests to run.**

The only exception to this would be if you are running Karma programmatically (not from the command line), in which case you might manually invoke `executor.schedule()` or emit your own `"run_start"` events.

Personally, I’ve begun to think that `config.autoWatch` and `config.singleRun` have an awkward relationship — which I think is demonstrated by setting them both to false.

***

**TL;DR**: based on the way Karma is designed, it looks like we almost always want `config.autoWatch` set to `true` — and we simply toggle `config.singleRun` as needed.

Does any of this help you?

**Comment by jlmakes** [comment]
Hey Mickael

Thanks for your contributions, let’s see if we can get to the bottom of this!

If you don't mind, I'd like to first share my understanding of the Karma source (as it pertains to `config.autoWatch`) — I think this will help us communicate about your problem.

Most of Karma is wrapped up into the `Server` found at `./lib/server.js`, and a search for `config.autoWatch` has three matches (all in the `_start` method):

<details>
<summary><i>Expand for code review</i></summary>

***

[First](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L168-L170) — It is used to initialize the Karma watcher after preprocessing completes:

```js
const afterPreprocess = () => {
  if (config.autoWatch) {
    this._injector.invoke(watcher.watch);
  }
  // other code...
};
```

This doesn’t really impact us, because we should be passing `watched: false` to the Karma config:

```js
files: [
  { pattern: 'test/**/*.spec.js', watched: false }
],
```

***

[Second](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L191-L201) — It is used for the first run of tests after all browsers are ready:

```js
this.on("browser_register", browser => {
  launcher.markCaptured(browser.id);

  if (launcher.areAllCaptured()) {
    this.emit("browsers_ready");

    if (config.autoWatch) {
      executor.schedule();
    }
  }
});

```

***

[Third](https://github.com/karma-runner/karma/blob/b8b8618b08acf14e3e4d435904029fe7a18134f3/lib/server.js#L324-L332) — It is used to bind `executor.schedule()` to the `"file_list_modified"` event:

```js
if (config.autoWatch) {
  this.on("file_list_modified", () => {
    this.log.debug("List of files has changed, trying to execute");
    if (config.restartOnFileChange) {
      socketServer.sockets.emit("stop");
    }
    executor.schedule();
  });
}
```

Both the Karma’s watcher and the `karma-rollup-preprocessor@next` watcher rely upon this event handler to re-run tests.

</details>

***

We can see that by setting `config.autoWatch` to `false`, we are telling Karma:

- Don‘t initialize a watcher
- Don‘t run tests after `karma start`
- Don’t re-run tests when files are modified

Which means your test suite will just hang: no tests will run, even if you modify files. You can _force_ your tests to run in this situation using `config.singleRun == true`, which under the hood emits the same `"run_start"` event that `executor.schedule()` does.

> ...it looks like my unit test suite never ends when autoWatch is set to false.

**Either`config.singleRun` or `config.autoWatch` must be `true` in order for your tests to run.**

The only exception to this would be if you are running Karma programmatically (not from the command line), in which case you might manually invoke `executor.schedule()` or emit your own `"run_start"` events.

Personally, I’ve begun to think that `config.autoWatch` and `config.singleRun` have an awkward relationship — which I think is demonstrated by setting them both to `false`.

***

**TL;DR**: based on the way Karma is designed, it looks like we almost always want `config.autoWatch` set to `true` — and we simply toggle `config.singleRun` as needed.

Does any of this help you?

**Comment by mjeanroy** [comment]
Hi,
> Thanks for your contributions, let’s see if we can get to the bottom of this!
As I said, this is PR may not be the most appropriate fix, so yeah let's see if we can do better!

> If config.autoWatch is false then we can assume that config.singleRun is true — which will kill the process (and any watcher spawned therein), so I'm not terribly put off by instantiating a watcher by default for simpler code.

In theory, you are right, but I think something is missing in the [`Watcher`](https://github.com/jlmakes/karma-rollup-preprocessor/blob/development/lib/Watcher.js) implementation. in fact the main problem is that watched file does not seem to be cleaned from the [internal map](https://github.com/jlmakes/karma-rollup-preprocessor/blob/2921c42efeb1836cd1ffa740936beec7a364985c/lib/Watcher.js#L9) used in the `Watcher` implementation.

Let me explain this:
- When the bundle is preprocessed, [the entry point and its dependencies are added to the internal `Watcher` map](https://github.com/jlmakes/karma-rollup-preprocessor/blob/2921c42efeb1836cd1ffa740936beec7a364985c/lib/index.js#L24) to be notified about changes (so far, so good).
- The [entry point and the dependencies are then added to a map](https://github.com/jlmakes/karma-rollup-preprocessor/blob/2921c42efeb1836cd1ffa740936beec7a364985c/lib/Watcher.js#L19) inside the `Watcher` implementation.
- These files are never cleared from the map, and I think this is the reason the process never ends (and if I removed that line, my problem is solved).

Do you know if we can listen to an event at the end of the process to clear the map? I think, it could solve all the problems here.

If you want to go deeper, just checkout this [PR in my project](https://github.com/mjeanroy/jasmine-utils/pull/330) here, update karma-rollup-preprocessor to the next version (just run `npm install && npm install --save-dev karma-rollup-preprocessor@next`) and run `npm test` (the process will never exit).

**Comment by jlmakes** [comment]
Thanks for the reproduction!

You are using Gulp to run Karma [programmatically](https://github.com/mjeanroy/jasmine-utils/blob/master/scripts/test/index.js#L100) — and because you’ve overwritten Karma’s `done()` callback to call Gulp’s `done()`, there is no `process.exit()` call (Karma’s [default](https://github.com/karma-runner/karma/blob/acdd2dc2ccdd054f0845455d762919f5f2dff561/lib/server.js#L73)) and so the Gulp child process for Karma lives on.

It's not clear to me why adding _this_ to the watcher doesn’t solve it:

```js
this.emitter.on("exit", () => {
  this.watch.close();
})
```

But I’ve now found a lot of issues on the web concerning issues stopping the Karma process, without resorting to `process.exit()`

**Comment by jlmakes** [comment]
Thanks for the reproduction!

You are using Gulp to run Karma [programmatically](https://github.com/mjeanroy/jasmine-utils/blob/master/scripts/test/index.js#L100) — and because you’ve overwritten Karma’s `done()` callback to call Gulp’s `done()`, there is no `process.exit()` call (Karma’s [default](https://github.com/karma-runner/karma/blob/acdd2dc2ccdd054f0845455d762919f5f2dff561/lib/server.js#L73)) and so the Gulp child process for Karma lives on.

I was hoping it’d be as simple as adding something like this to the watcher:

```js
process.on("exit", () => {
  this.watch.close();
})
```

But that doesn’t seem to work — and I’ve now found a lot of issues on the web concerning issues stopping the Karma process, without resorting to `process.exit()`

**Comment by jlmakes** [comment]
Thanks for the reproduction!

You are using Gulp to run Karma [programmatically](https://github.com/mjeanroy/jasmine-utils/blob/master/scripts/test/index.js#L100) — and because you’ve overwritten Karma’s `done()` callback to call Gulp’s `done()`, there is no `process.exit()` call (Karma’s [default](https://github.com/karma-runner/karma/blob/acdd2dc2ccdd054f0845455d762919f5f2dff561/lib/server.js#L73)) and so the Gulp child process for Karma lives on.

I was hoping it’d be as simple as adding something like this to the watcher:

```js
this.emitter.on("exit", () => {
  this.watch.close();
})
```

But that seems to make things worse — and I’ve now found a lot of issues on the web concerning issues stopping the Karma process, without resorting to `process.exit()`

Even if this is an upstream bug, I’m warming up to your PR! 🤣

**Comment by mjeanroy** [comment]
@jlmakes What do you think about this PR? Do you think it should be included in next 7.0.0 version or would you like to go deeper with another solution?

**Comment by jlmakes** [comment]
Well my experiments with an event based solution seem to suggest an upstream oddity with Karma, and I think your solution is clear and practical.

I’ve been thinking about it, and instead of disabling the watcher when `config.autoWatch` is `false`, I’m thinking of disabling the watcher when `config.singleRun` is `true`.

It’s a subtle distinction, but I think it will handle programmatic use cases better for when `config.autoWatch` and `config.singleRun` are both `false`.

I’ll make the change and ping you later today when I release `7.0.0-rc.3`

**Comment by jlmakes** [comment]
Well my experiments with an event based solution seem to suggest an upstream oddity with Karma, and I think your solution is clear and practical.

I’ve been thinking about it, and I’m trying to decide whether to disable the watcher when `config.autoWatch` is `false`, or when `config.singleRun` is `true` — because in both cases, the watcher could/should not run.

It’s a subtle distinction, but I’m wondering if there are some programmatic edge cases where it might make a difference. In any case, I will implement one or the other and ping you tomorrow today when I release `7.0.0-rc.3`

**Comment by jlmakes** [comment]
Yes, I’ll be including this fix in the upcoming release.

My experiments with an event based solution seem to suggest an upstream oddity with Karma, and I think your solution is clear and practical. Since it turns out I'm pulling in a few more properties from Karma `config`, I'm refactoring the dependency injection; `7.0.0` will be published this week.

**Comment by jlmakes** [comment]
@mjeanroy I made sure the fix landed in `7.0.0` — I tested with your project, and everything seems OK!
